### DISCUSSION TITLE
How to get the current sequence number?

### DISCUSSION THREAD

**Discussion by clmnin** [geldata/gel#2750] [Q&A] [answered]
Two new functions were added to manipulate a sequence - [sequence_next](https://www.edgedb.com/docs/edgeql/funcops/sequence#function::std::sequence_next) and [sequence_reset](https://www.edgedb.com/docs/edgeql/funcops/sequence#function::std::sequence_reset). 

But I can't  find the documentation for getting the current sequence value.

**Comment by elprans** [comment, marked as answer]
This is because it's unimplemented.  Getting the current value of a sequence is quirky in Postgtres: it'll only work if you used `sequence_next()` in the same session, otherwise it'll fail.  In other words, there isn't a universal "current sequence number", only "the last value returned by sequence_next() in this session".

**Reply by clmnin** [reply]
My use case was where I need to get `sequence_current` in the same session

```
SELECT (
    INSERT SMS {
        body := "Your order #" ++ sequence_current(INTROSPECT SaleOrderSequence) ++ "was placed successfully",
        so := (
            INSERT SaleOrder {
                customer := (SELECT User FILTER .name = <str>$customer_name), 
                merchant := (SELECT User FILTER .name = <str>$merchant_name), 
                lines := {
                    (INSERT OrderLine { product := (SELECT Product FILTER .name = <str>$product_name), qty := <int32>$order_qty})
                }
            }
        )
    }
) {
    body,
    so : {
        customer,
        merchant,
        lines : {
            product,
            qty
        }
    }
};
```

**Questions**
1. I need to get the order number (SaleOrderSequence) for the message as I INSERT a new Sale Order. Is this possible now?
2. Currently we are restricted from using `USING SQL $$` in a EdgeQL function. Will this always be barred from EdgeDB users?

**Extra stuff**

* An old dba.stackexchange questions answers the question of "getting the current sequence value". - [Link](https://dba.stackexchange.com/a/3284)
* For others who may come across this problem
    * pg has a `currval` function to get the current value of a sequence. But an error is reported if nextval has never been called for this sequence in this session) - [Doc](https://www.postgresql.org/docs/13/functions-sequence.html)
    ```
    currval ( regclass ) → bigint
    Returns the value most recently obtained by nextval for this sequence in the current session. (An error is reported
    if nextval has never been called for this sequence in this session.) Because this is returning a session-local value,
    it gives a predictable answer whether or not other sessions have executed nextval since the current session did.
    
    This function requires USAGE or SELECT privilege on the sequence.
    ```

**Reply by elprans** [reply]
You wouldn't need `sequence_current` if you rewrite your query like so:

```
WITH
   NewOrder := (
       INSERT SaleOrder {
           customer := (SELECT User FILTER .name = <str>$customer_name), 
           merchant := (SELECT User FILTER .name = <str>$merchant_name), 
           lines := {
               (INSERT OrderLine { product := (SELECT Product FILTER .name = <str>$product_name), qty := <int32>$order_qty})
           }
       }
   )
SELECT (
    INSERT SMS {
        body := "Your order #" ++ NewOrder.number ++ "was placed successfully",
        so := NewOrder,
    }
) {
    body,
    so : {
        customer,
        merchant,
        lines : {
            product,
            qty
        }
    }
};
```

I'm assuming above that the sequence property is called "number", if not, adjust as necessary.

**Reply by clmnin** [reply]
Thank you. Please answer this question too (much more of a roadmap question)

1. Currently we are restricted from using `USING SQL $$` in a EdgeQL function. Will this always be barred from EdgeDB users?

**Reply by elprans** [reply]
SQL is an implementation detail in EdgeDB, and allowing direct access to it from EdgeQL risks all kinds of breakage and undefined behavior.  That said, we are planning to add a formal extension system where you'd be able to define custom scalar types and accompanying functions, possibly with SQL (exposing PostGIS functionality is a good example).
